api/include/opentelemetry/metrics/async_instruments.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "instrument.h"
+#include "observer_result.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace metrics
+{
+
+class IntValueObserver : public AsynchronousInstrument
+{
+
+public:
+  IntValueObserver() = default;
+
+  IntValueObserver(nostd::string_view name,
+                   nostd::string_view description,
+                   nostd::string_view unit,
+                   bool enabled,
+                   void (*callback)(ObserverResult))
+  {}
+
+  /*
+   * Updates the instruments aggregator with the new value. The labels should
+   * contain the keys and values to be associated with this value.
+   *
+   * @param value is the numerical representation of the metric being captured
+   * @param labels the set of labels, as key-value pairs
+   */
+  virtual void observe(int value, const trace::KeyValueIterable &labels) {}
+};
+
+class DoubleValueObserver : public AsynchronousInstrument
```

Comment:
It looks like there are a lot of these that only differ based on the type of value that is accepted by the `observe` method. Could this be generalized through templating or macros?

I'm also fine leaving as is and updating if/when there is reason to (e.g reduction of duplicated methods added to each of the instruments).